### PR TITLE
TransferDetails Services

### LINKED ISSUE #1110: TransferDetails User API
<This is intended as a guide when thinking about or creating issues for go-centrifuge>

This new extension should be implemented with a new set of custom attributes and allow for transfer details to be added to a Document.

### PR DESCRIPTION
**Pull request by charleenfei**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Closes #1110 

- nolint on transferdetails/funding.go/userapi types.go files because of special needs for parsing due to custom attributes (temporary solution)
- separates out generic attribute utils into
- services and testing
- some work on migrating funding extension to userAPI format, but this will be finished in a separate PR

opening another PR for the userapi portion due to size of this PR.

### CODE REVIEW

File: extensions/transferdetails/converters.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package transferdetails
```
Comment (charleenfei) [inline extensions/transferdetails/converters.go]: this file will be removed.

File: extensions/funding/converters.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package funding
+
+import (
+	"reflect"
+
+	"github.com/centrifuge/go-centrifuge/protobufs/gen/go/funding"
+)
+
+// TODO: generic? Create something like ServiceRegistry for Custom Attributes...
+func (f *Data) initFundingFromData(data *funpb.FundingData) {
```
Comment (manuelpolzhofer) [inline extensions/funding/converters.go]: I think this method could be generic

File: extensions/transferdetails/service.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+			}
+			c = append(c, did)
+		}
+	}
+
+	return c, nil
+}
+
+func (s service) DeriveFromPayload(ctx context.Context, req CreateTransferDetailRequest, identifier []byte) (documents.Model, error) {
+	model, err := s.GetCurrentVersion(ctx, identifier)
```
Comment (manuelpolzhofer) [inline extensions/transferdetails/service.go]: This method could be more generic.

Where you pass in `transfersFieldKey, transfersLabel`  and the `dids`

File: extensions/transferdetails/service.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	idx, err := extensions.FindAttributeSetIDX(model, transferID, transfersLabel, transferIDLabel, transfersFieldKey)
+	if err != nil {
+		return nil, err
+	}
+	return s.deriveTransferData(model, idx)
+}
+
+// TODO: generic?
+func (s service) deriveTransferData(model documents.Model, idx string) (*TransferDetailData, error) {
+	data := new(TransferDetailData)
```
Comment (manuelpolzhofer) [inline extensions/transferdetails/service.go]: Same here:

The only individual part is:`data := new(TransferDetailData)`

File: extensions/transferdetails/service.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+	return &TransferDetailResponse{
+		Header: h,
+		Data:   data,
+	}, nil
+
+}
+
+// DeriveTransfersListResponse returns a transfers list
+func (s service) DeriveTransferListResponse(ctx context.Context, model documents.Model) (*TransferDetailListResponse, error) {
+	response := new(TransferDetailListResponse)
```
Comment (manuelpolzhofer) [inline extensions/transferdetails/service.go]: Same here:

We only need an additional parameter for the struct to be generic. like ` TransferDetailData` or `Funding`

File: extensions/transferdetails/service.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+		}
+
+	}
+	return data, nil
+}
+
+// DeriveTransferResponse returns create response from the added transfer detail
+func (s service) DeriveTransferResponse(ctx context.Context, model documents.Model, transferID string) (*TransferDetailResponse, error) {
+	idx, err := extensions.FindAttributeSetIDX(model, transferID, transfersLabel, transferIDLabel, transfersFieldKey)
```
Comment (manuelpolzhofer) [inline extensions/transferdetails/service.go]: same here:

the method could be moved to the extension package together with an extra parameter for the struct.

this method would only call the generic method and afterwards return the `&TransferDetailResponse{...}`

File: extensions/funding/converters.go (hunk 2 above)
Comment (charleenfei) [reply extensions/funding/converters.go]: this whole file should be removed when we get rid of the protobuf :)

**Comment by charleenfei** [comment]
totally agree @manuelpolzhofer -- I think it should be part of the rethinking when we migrate the fundingAPI also over to the new userAPI structure.

File: extensions/transferdetails/service.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+	"github.com/centrifuge/go-centrifuge/identity"
+	logging "github.com/ipfs/go-log"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+	documents.Service
+
+	// DeriveFromUpdatePayload derives Funding from clientUpdatePayload
+	DeriveFromUpdatePayload(ctx context.Context, req UpdateTransferDetailRequest, identifier []byte) (documents.Model, error)
```
Comment (vedhavyas) [inline extensions/transferdetails/service.go]: Why not make the identifier part of the `req` ?

File: extensions/transferdetails/service.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/extensions"
+	"github.com/centrifuge/go-centrifuge/identity"
+	logging "github.com/ipfs/go-log"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+	documents.Service
```
Comment (vedhavyas) [inline extensions/transferdetails/service.go]: Lets not embed this, let keep it as a field instead

File: extensions/transferdetails/service.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	return model, nil
+}
+
+// DeriveFromUpdatePayload derives Funding from clientUpdatePayload
+func (s service) DeriveFromUpdatePayload(ctx context.Context, req UpdateTransferDetailRequest, identifier []byte) (documents.Model, error) {
+	model, err := s.GetCurrentVersion(ctx, identifier)
+	if err != nil {
+		transfersLog.Error(err)
```
Comment (vedhavyas) [inline extensions/transferdetails/service.go]:
```suggestion
		log.Error(err)
```

Comment (vedhavyas) [reply extensions/transferdetails/service.go]: We will have only one log instance for this entire package. So calling logger should be enough IMO

File: httpapi/userapi/transfer_details_api.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package userapi
+
+import (
+	logging "github.com/ipfs/go-log"
+)
+
+var transfersLog = logging.Logger("tranfers-api")
```
Comment (vedhavyas) [inline httpapi/userapi/transfer_details_api.go]:
```suggestion
var log = logging.Logger("user-api")
```

Comment (vedhavyas) [reply httpapi/userapi/transfer_details_api.go]: TransferAPIService in `transferdetails` package will have its own logger and UserAPI logger is used for all USerAPI logs

File: httpapi/userapi/types.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+// TransferDetailResponse
+type TransferDetailResponse struct {
+	Header *documentpb.ResponseHeader          `json:"header"`
+	Data   *transferdetails.TransferDetailData `json:"data"`
+}
+
+// TransferDetailListResponse
+type TransferDetailListResponse struct {
+	Header *documentpb.ResponseHeader            `json:"header"`
```
Comment (vedhavyas) [inline httpapi/userapi/types.go]: We want to get rid of protobufs. Use the CoreAPI ResponseHeader instead

File: httpapi/userapi/types.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+//CreateTransferDetailRequest
+type CreateTransferDetailRequest struct {
+	Identifier string                              `json:"identifier" swaggertype:"primitive,string"`
+	Data       *transferdetails.TransferDetailData `json:"data"`
+}
+
+// UpdateTransferDetailRequest
+type UpdateTransferDetailRequest struct {
+	Identifier string                              `json:"identifier" swaggertype:"primitive,string"`
+	TransferId string                              `json:"transfer_id" swaggertype:"primitive,string"`
```
Comment (vedhavyas) [inline httpapi/userapi/types.go]:
```suggestion
	TransferID string                              `json:"transfer_id" swaggertype:"primitive,string"`
```

File: httpapi/userapi/types.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+package userapi
+
+import (
+	"github.com/centrifuge/go-centrifuge/extensions/transferdetails"
+	"github.com/centrifuge/go-centrifuge/protobufs/gen/go/document"
+)
+
+// TransferDetailData is the default transfer details extension schema
+type TransferDetailData struct {
+	TransferId          string `json:"transfer_id,omitempty" attr:"bytes"`
```
Comment (vedhavyas) [inline httpapi/userapi/types.go]:
```suggestion
	TransferID          string `json:"transfer_id,omitempty" attr:"bytes"`
```

**Comment by charleenfei** [comment]
closing this to open smaller PRs

**Pull request closed without merging**